File: packages/icp/src/contract/coding-agent-contract.ts
/**
 * Coding agent input/output contract (LAT-174).
 *
 * Defines the canonical shape of the coding agent's contract: what the agent
 * receives as input (repo, branch, targetBranch, code diff, build command,
 * test command) and what it produces as output (diff, buildStatus, testResults,
 * lintResults, coverage).
 *
 * Evidence requirements are drawn from LAT-8 (Define QA/review evidence
 * workflow), which mandates that every coding-agent run produces structured
 * evidence for:
 *   - Acceptance criteria verification
 *   - Test results
 *   - Files changed
 *   - Risks
 *   - Regressions
 *   - Security/architecture concerns
 *   - Final recommendation
 *
 * This contract is a **pure type layer** — it does not call any adapters,
 * run commands, or perform I/O. Adapters (the command provider, MCP server,
 * Claude Code harness) normalise their own outputs into these types so the
 * rest of the stack works against a stable surface.
 *
 * LAT-174 non-goals: agent-specific output formats (adapters normalise).
 */

/* ------------------------------------------------------------------ */
/* Schema version                                                    */
/* ------------------------------------------------------------------ */

/**
 * Schema version for the coding agent contract. Bumped on breaking changes.
 * The runner and adapters use this to verify compatibility.
 */
export const CODING_AGENT_CONTRACT_SCHEMA_VERSION = "1.0.0";

/* ------------------------------------------------------------------ */
/* Input types                                                       */
/* ------------------------------------------------------------------ */

/**
 * The canonical input a coding agent receives before it starts work.
 *
 * Mirrors the LAT-8 evidence requirements: the agent needs enough
 * context to verify acceptance criteria, run build/tests/lint, and
 * produce a structured output with all required evidence fields.
 *
 * LAT-174 input fields (per the issue acceptance criteria):
 *   - repo:         the target repository (`owner/name`)
 *   - branch:       the source branch the agent will work on
 *   - targetBranch: the base branch for the PR (typically `main`)
 *   - codeDiff:     the code diff or patch the agent should apply/return
 *   - buildCommand:  command to build the project
 *   - testCommand:   command to run tests
 */
export interface CodingAgentInput {
  /**
   * Target repository in `owner/name` form (e.g. `BenjaminDElliott/latentspacelabs`).
   * The agent opens/closes PRs against this repo.
   */
  repo: string;

  /**
   * Source branch the agent will create/modify.
   * Conventionally follows the pattern `lat-<issueNumber>-<slug>`.
   */
  branch: string;

  /**
   * Target (base) branch for the PR. Typically `main` or `develop`.
   * The agent opens a PR from `branch` → `targetBranch`.
   */
  targetBranch: string;

  /**
   * Code diff or patch the agent should produce (or apply on top of the
   * working tree). Can be empty if the agent generates code from scratch.
   * Stored as a unified diff string (git-diff format) for structured
   * consumption by downstream reviewers.
   */
  codeDiff?: string;

  /**
   * Shell command to build the project. Executed inside the agent's
   * sandbox/worktree. Must exit 0 on success.
   *
   * Example: `npm run build` or `make -j4`
   */
  buildCommand: string;

  /**
   * Shell command to run tests. Executed inside the agent's sandbox/
   * worktree. Must exit 0 on success.
   *
   * Example: `npm test` or `pytest tests/`
   */
  testCommand: string;

  /**
   * Optional shell command to run linting. If absent, linting is skipped
   * in the output (lintResults.status defaults to "not_run").
   *
   * Example: `npm run lint` or `flake8 src/`
   */
  lintCommand?: string;

  /**
   * Optional shell command to measure test coverage. If absent, coverage
   * is omitted from the output (coverage.totalPercentage defaults to null).
   *
   * Example: `npm run coverage` or `pytest --cov=src`
   */
  coverageCommand?: string;

  /**
   * Acceptance criteria the agent should verify in its output.
   * Mirrors LAT-8's acceptance criteria verification requirement.
   * Each criterion should be a single testable statement.
   */
  acceptanceCriteria?: ReadonlyArray<string>;

  /**
   * Files in scope for the task. Helps the agent avoid editing
   * irrelevant files. Mirrors LAT-8's files changed requirement.
   */
  filesInScope?: ReadonlyArray<string>;

  /**
   * Files explicitly excluded from the task.
   */
  filesForbidden?: ReadonlyArray<string>;

  /**
   * Non-goals / out-of-scope notes. The agent should not touch these.
   */
  nonGoals?: ReadonlyArray<string>;

  /**
   * Guardrails / rules the agent must follow (e.g. ADR-0008, ADR-0013).
   */
  guardrails?: ReadonlyArray<string>;

  /**
   * Numeric ADR-0009 budget cap for the run. The agent should surface
   * cost band evidence in its output.
   */
  budgetCapUsd?: number | null;

  /**
   * ADR-0008 autonomy level for this run (L1–L4). Affects what the
   * agent is allowed to do (propose, approve, auto-merge).
   */
  autonomyLevel?: AutonomyLevel;

  /**
   * Stable run identifier for correlation with the ICP runner.
   */
  runId?: string;

  /**
   * Ticket title from Linear. Used as the PR title prefix.
   */
  ticketTitle?: string;
}

/* ------------------------------------------------------------------ */
/* Output types                                                      */
/* ------------------------------------------------------------------ */

/**
 * The canonical output a coding agent produces after completing its task.
 *
 * Every field maps to an LAT-8 evidence requirement:
 *   - diff              → files changed
 *   - buildStatus       → build verification
 *   - testResults       → test results
 *   - lintResults       → lint verification
 *   - coverage          → coverage evidence
 *
 * LAT-174 output fields (per the issue acceptance criteria):
 *   - diff:              the code diff / patch produced
 *   - buildStatus:       whether the build passed
 *   - testResults:       structured test pass/fail results
 *   - lintResults:       lint pass/fail results
 *   - coverage:          test coverage metrics
 *
 * Plus LAT-8 evidence fields:
 *   - acceptanceCriteriaVerified
 *   - risks
 *   - regressions
 *   - securityOrArchitectureConcerns
 *   - recommendation
 */
export interface CodingAgentOutput {
  /**
   * Schema version. Consumers use this to branch on schema without
   * re-reading the contract.
   */
  schemaVersion: typeof CODING_AGENT_CONTRACT_SCHEMA_VERSION;

  /**
   * The code diff produced by the agent. Stored as a unified diff
   * (git-diff format). Mirrors LAT-8's "files changed" evidence.
   * Empty string when no code was produced (e.g. a documentation-only
   * ticket or a refused run).
   */
  diff: string;

  /**
   * Build result. Mirrors LAT-8's build verification evidence.
   */
  buildStatus: BuildStatus;

  /**
   * Test results. Mirrors LAT-8's test results evidence.
   * Contains structured per-test pass/fail data.
   */
  testResults: TestResults;

  /**
   * Lint results. Mirrors LAT-8's lint verification evidence.
   * Empty when no lint command was provided.
   */
  lintResults: LintResults;

  /**
   * Coverage metrics. Mirrors LAT-8's coverage evidence.
   * Null when no coverage command was provided.
   */
  coverage: CoverageMetrics | null;

  /**
   * LAT-8 evidence: acceptance criteria verification. Each criterion
   * is verified as `passed`, `failed`, `partial`, or `skipped`.
   */
  acceptanceCriteriaVerified: ReadonlyArray<AcceptanceCriterionResult>;

  /**
   * LAT-8 evidence: identified risks during the run.
   * One-line risk descriptions. Empty array means no risks identified.
   */
  risks: ReadonlyArray<string>;

  /**
   * LAT-8 evidence: potential regressions from the changes.
   * One-line regression descriptions. Empty array means no regressions identified.
   */
  regressions: ReadonlyArray<string>;

  /**
   * LAT-8 evidence: security and architecture concerns flagged by the agent.
   * One-line concern descriptions. Empty array means no concerns flagged.
   */
  securityOrArchitectureConcerns: ReadonlyArray<string>;

  /**
   * LAT-8 evidence: final recommendation for the reviewer/merger.
   * Mirrors the ADR-0007 recommendation ladder: `approve`, `approve-with-nits`,
   * `request-changes`, `block-merge`, or `needs-human`.
   */
  recommendation: Recommendation;

  /**
   * One-line summary of what the agent did. Used in Linear write-back.
   */
  summary: string;

  /**
   * Pointers to PR, commit, and branch for correlation with the ICP runner.
   */
  prUrl?: string | null;
  prBranch?: string | null;
  commitSha?: string | null;

  /**
   * Free-form notes from the agent (sanitised of secrets).
   */
  notes?: ReadonlyArray<string>;
}

/* ------------------------------------------------------------------ */
/* Build result                                                       */
/* ------------------------------------------------------------------ */

/**
 * Build status from the coding agent's output.
 * Maps directly to the exit code semantics.
 */
export type BuildStatus = {
  /** Overall status. */
  status: "passed" | "failed" | "not_run";
  /** Exit code from the build command (null if not_run). */
  exitCode: number | null;
  /** Stdout from the build command, sanitised (secrets redacted). */
  stdout: string;
  /** Stderr from the build command, sanitised (secrets redacted). */
  stderr: string;
  /** Duration in milliseconds. */
  durationMs: number;
};

/* ------------------------------------------------------------------ */
/* Test results                                                       */
/* ------------------------------------------------------------------ */

/**
 * Overall test outcome.
 */
export type TestOutcome = "passed" | "failed" | "skipped" | "not_run";

/**
 * Structured test results from the coding agent's output.
 * Mirrors LAT-8's "test results" evidence requirement.
 */
export interface TestResults {
  /** Overall outcome. */
  outcome: TestOutcome;
  /** Total test count (including passed, failed, skipped). */
  totalTests: number;
  /** Number of tests that passed. */
  passed: number;
  /** Number of tests that failed. */
  failed: number;
  /** Number of tests that were skipped. */
  skipped: number;
  /** Number of tests that errored (crashed during execution). */
  errors: number;
  /** Duration in milliseconds. */
  durationMs: number;
  /** Individual test results. Empty when outcome is "not_run". */
  tests: ReadonlyArray<TestCaseResult>;
}

/**
 * Result for a single test case.
 */
export interface TestCaseResult {
  name: string;
  /** Suite/module this test belongs to. */
  suite?: string;
  status: "passed" | "failed" | "skipped" | "errored";
  /** Duration in milliseconds for this test. */
  durationMs: number;
  /** Error message if failed/errored. */
  errorMessage?: string;
  /** Stack trace if failed/errored, sanitised. */
  stackTrace?: string;
}

/* ------------------------------------------------------------------ */
/* Lint results                                                       */
/* ------------------------------------------------------------------ */

/**
 * Lint result status.
 */
export type LintStatus = "passed" | "failed" | "not_run";

/**
 * Structured lint results from the coding agent's output.
 * Mirrors LAT-8's lint verification evidence.
 */
export interface LintResults {
  status: LintStatus;
  /** Total number of lint violations found. */
  totalViolations: number;
  /** Number of errors (must-fix). */
  errors: number;
  /** Number of warnings (should-fix). */
  warnings: number;
  /** Number of infos/notices (cosmetic). */
  infos: number;
  /** Duration in milliseconds. */
  durationMs: number;
  /** Individual lint violations. Empty when status is "not_run". */
  violations: ReadonlyArray<LintViolation>;
}

/**
 * A single lint violation.
 */
export interface LintViolation {
  /** File path where the violation was found. */
  file: string;
  /** Line number (1-based). */
  line: number;
  /** Column number (1-based). */
  column?: number;
  /** Severity level. */
  severity: "error" | "warning" | "info";
  /** Rule identifier. */
  rule: string;
  /** Human-readable message. */
  message: string;
}

/* ------------------------------------------------------------------ */
/* Coverage metrics                                                   */
/* ------------------------------------------------------------------ */

/**
 * Coverage metrics from the coding agent's output.
 * Mirrors LAT-8's coverage evidence requirement.
 */
export interface CoverageMetrics {
  /** Overall percentage of covered lines (0–100). Null if not computed. */
  totalPercentage: number | null;
  /** Percentage of covered branches (0–100). Null if not computed. */
  branchPercentage: number | null;
  /** Percentage of covered functions (0–100). Null if not computed. */
  functionPercentage: number | null;
  /** Percentage of covered statements (0–100). Null if not computed. */
  statementPercentage: number | null;
  /** Duration in milliseconds. */
  durationMs: number;
  /**
   * Per-file coverage summary. Ordered by file path.
   */
  files: ReadonlyArray<CoverageFileSummary>;
}

/**
 * Coverage for a single file.
 */
export interface CoverageFileSummary {
  /** File path. */
  file: string;
  /** Lines covered / total lines (percentage). Null if not available. */
  lineCoverage: number | null;
  /** Branches covered / total branches. Null if not available. */
  branchCoverage: number | null;
  /** Functions covered / total functions. Null if not available. */
  functionCoverage: number | null;
  /** Statements covered / total statements. Null if not available. */
  statementCoverage: number | null;
}

/* ------------------------------------------------------------------ */
/* Acceptance criterion result                                        */
/* ------------------------------------------------------------------ */

/**
 * Verification result for a single acceptance criterion.
 * Mirrors LAT-8's acceptance criteria verification requirement.
 */
export interface AcceptanceCriterionResult {
  /** The criterion text (verbatim from input). */
  criterion: string;
  /** Verification result. */
  status: "passed" | "failed" | "partial" | "skipped";
  /** Evidence or explanation. */
  evidence: string;
}

/* ------------------------------------------------------------------ */
/* Recommendation                                                     */
/* ------------------------------------------------------------------ */

/**
 * Recommendation ladder from ADR-0007 (LAT-26 §6.2).
 * Used by LAT-8 evidence workflow and the review agent.
 */
export type Recommendation =
  | "approve"
  | "approve-with-nits"
  | "request-changes"
  | "block-merge"
  | "needs-human";

/* ------------------------------------------------------------------ */
/* Autonomy level                                                     */
/* ------------------------------------------------------------------ */

/**
 * Autonomy notation per ADR-0008 (L0–L5).
 * Re-exported for convenience so this module is self-contained.
 */
export type AutonomyLevel =
  | "L1-read-only"
  | "L2-propose"
  | "L3-with-approval"
  | "L4-autonomous";

File: packages/icp/src/contract/coding_agent_contract.py
"""
Coding agent input/output contract (LAT-174) — Python types.

Defines the canonical shape of the coding agent's contract: what the agent
receives as input (repo, branch, target_branch, code diff, build command,
test command) and what it produces as output (diff, build_status, test_results,
lint_results, coverage).

Evidence requirements are drawn from LAT-8 (Define QA/review evidence
workflow), which mandates that every coding-agent run produces structured
evidence for:
  - Acceptance criteria verification
  - Test results
  - Files changed
  - Risks
  - Regressions
  - Security/architecture concerns
  - Final recommendation

This module is a pure type layer — no I/O, no side effects.

LAT-174 non-goals: agent-specific output formats (adapters normalise).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


# ------------------------------------------------------------------ #
# Schema version                                                     #
# ------------------------------------------------------------------ #

CODING_AGENT_CONTRACT_SCHEMA_VERSION = "1.0.0"


# ------------------------------------------------------------------ #
# Enums                                                              #
# ------------------------------------------------------------------ #


class AutonomyLevel(str, Enum):
    """ADR-0008 autonomy notation (L1–L4)."""

    L1_READ_ONLY = "L1-read-only"
    L2_PROPOSE = "L2-propose"
    L3_WITH_APPROVAL = "L3-with-approval"
    L4_AUTONOMOUS = "L4-autonomous"


class BuildStatusValue(str, Enum):
    """Build status outcome."""

    PASSED = "passed"
    FAILED = "failed"
    NOT_RUN = "not_run"


class TestOutcome(str, Enum):
    """Overall test outcome."""

    PASSED = "passed"
    FAILED = "failed"
    SKIPPED = "skipped"
    NOT_RUN = "not_run"


class TestCaseStatus(str, Enum):
    """Result for a single test case."""

    PASSED = "passed"
    FAILED = "failed"
    SKIPPED = "skipped"
    ERRORED = "errored"


class LintStatusValue(str, Enum):
    """Lint result status."""

    PASSED = "passed"
    FAILED = "failed"
    NOT_RUN = "not_run"


class LintSeverity(str, Enum):
    """Lint violation severity."""

    ERROR = "error"
    WARNING = "warning"
    INFO = "info"


class AcceptanceCriterionStatus(str, Enum):
    """Verification result for a single acceptance criterion."""

    PASSED = "passed"
    FAILED = "failed"
    PARTIAL = "partial"
    SKIPPED = "skipped"


class Recommendation(str, Enum):
    """ADR-0007 recommendation ladder (LAT-26 §6.2)."""

    APPROVE = "approve"
    APPROVE_WITH_NITS = "approve-with-nits"
    REQUEST_CHANGES = "request-changes"
    BLOCK_MERGE = "block-merge"
    NEEDS_HUMAN = "needs-human"


# ------------------------------------------------------------------ #
# Input types                                                        #
# ------------------------------------------------------------------ #


@dataclass
class CodingAgentInput:
    """
    The canonical input a coding agent receives before it starts work.

    LAT-174 input fields:
      - repo:            target repository (owner/name)
      - branch:          source branch the agent will work on
      - target_branch:   base branch for the PR (typically main)
      - code_diff:       code diff/patch the agent should produce
      - build_command:   command to build the project
      - test_command:    command to run tests
      - lint_command:    optional command to run linting
      - coverage_command: optional command to measure coverage
      - acceptance_criteria: testable statements to verify
      - files_in_scope:   files the agent should touch
      - files_forbidden:  files explicitly excluded
      - non_goals:        out-of-scope notes
      - guardrails:       rules the agent must follow
      - budget_cap_usd:   ADR-0009 budget cap
      - autonomy_level:   ADR-0008 autonomy level (L1–L4)
      - run_id:           stable run identifier
      - ticket_title:     ticket title for PR title prefix
    """

    repo: str
    branch: str
    target_branch: str
    code_diff: str = ""
    build_command: str = ""
    test_command: str = ""
    lint_command: str = ""
    coverage_command: str = ""
    acceptance_criteria: list[str] = field(default_factory=list)
    files_in_scope: list[str] = field(default_factory=list)
    files_forbidden: list[str] = field(default_factory=list)
    non_goals: list[str] = field(default_factory=list)
    guardrails: list[str] = field(default_factory=list)
    budget_cap_usd: float | None = None
    autonomy_level: AutonomyLevel | None = None
    run_id: str = ""
    ticket_title: str = ""


# ------------------------------------------------------------------ #
# Build result                                                       #
# ------------------------------------------------------------------ #


@dataclass
class BuildStatus:
    """Build result from the coding agent's output."""

    status: BuildStatusValue = BuildStatusValue.NOT_RUN
    exit_code: int | None = None
    stdout: str = ""
    stderr: str = ""
    duration_ms: int = 0


# ------------------------------------------------------------------ #
# Test results                                                       #
# ------------------------------------------------------------------ #


@dataclass
class TestCaseResult:
    """Result for a single test case."""

    name: str
    suite: str = ""
    status: TestCaseStatus = TestCaseStatus.PASSED
    duration_ms: int = 0
    error_message: str = ""
    stack_trace: str = ""


@dataclass
class TestResults:
    """Structured test results."""

    outcome: TestOutcome = TestOutcome.NOT_RUN
    total_tests: int = 0
    passed: int = 0
    failed: int = 0
    skipped: int = 0
    errors: int = 0
    duration_ms: int = 0
    tests: list[TestCaseResult] = field(default_factory=list)


# ------------------------------------------------------------------ #
# Lint results                                                       #
# ------------------------------------------------------------------ #


@dataclass
class LintViolation:
    """A single lint violation."""

    file: str
    line: int
    column: int = 1
    severity: LintSeverity = LintSeverity.WARNING
    rule: str = ""
    message: str = ""


@dataclass
class LintResults:
    """Structured lint results."""

    status: LintStatusValue = LintStatusValue.NOT_RUN
    total_violations: int = 0
    errors: int = 0
    warnings: int = 0
    infos: int = 0
    duration_ms: int = 0
    violations: list[LintViolation] = field(default_factory=list)


# ------------------------------------------------------------------ #
# Coverage metrics                                                   #
# ------------------------------------------------------------------ #


@dataclass
class CoverageFileSummary:
    """Coverage for a single file."""

    file: str
    line_coverage: float | None = None
    branch_coverage: float | None = None
    function_coverage: float | None = None
    statement_coverage: float | None = None


@dataclass
class CoverageMetrics:
    """Coverage metrics."""

    total_percentage: float | None = None
    branch_percentage: float | None = None
    function_percentage: float | None = None
    statement_percentage: float | None = None
    duration_ms: int = 0
    files: list[CoverageFileSummary] = field(default_factory=list)


# ------------------------------------------------------------------ #
# Acceptance criterion result                                        #
# ------------------------------------------------------------------ #


@dataclass
class AcceptanceCriterionResult:
    """Verification result for a single acceptance criterion."""

    criterion: str
    status: AcceptanceCriterionStatus
    evidence: str = ""


# ------------------------------------------------------------------ #
# Output types                                                       #
# ------------------------------------------------------------------ #


@dataclass
class CodingAgentOutput:
    """
    The canonical output a coding agent produces after completing its task.

    LAT-174 output fields:
      - diff:                 code diff/patch produced
      - build_status:         build result
      - test_results:         structured test results
      - lint_results:         lint results
      - coverage:             coverage metrics (nullable)
      - acceptance_criteria_verified: LAT-8 evidence
      - risks:                LAT-8 evidence
      - regressions:          LAT-8 evidence
      - security_or_architecture_concerns: LAT-8 evidence
      - recommendation:       LAT-8 evidence (ADR-0007)
      - summary:              one-line summary
      - pr_url:               PR URL
      - pr_branch:            PR branch
      - commit_sha:           commit SHA
      - notes:                free-form agent notes
    """

    schema_version: str = CODING_AGENT_CONTRACT_SCHEMA_VERSION
    diff: str = ""
    build_status: BuildStatus = field(default_factory=BuildStatus)
    test_results: TestResults = field(default_factory=TestResults)
    lint_results: LintResults = field(default_factory=LintResults)
    coverage: CoverageMetrics | None = None
    acceptance_criteria_verified: list[AcceptanceCriterionResult] = field(
        default_factory=list
    )
    risks: list[str] = field(default_factory=list)
    regressions: list[str] = field(default_factory=list)
    security_or_architecture_concerns: list[str] = field(default_factory=list)
    recommendation: Recommendation = Recommendation.NEEDS_HUMAN
    summary: str = ""
    pr_url: str | None = None
    pr_branch: str | None = None
    commit_sha: str | None = None
    notes: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        if self.schema_version != CODING_AGENT_CONTRACT_SCHEMA_VERSION:
            raise ValueError(
                f"schema_version must be {CODING_AGENT_CONTRACT_SCHEMA_VERSION!r}, "
                f"got {self.schema_version!r}"
            )


# ------------------------------------------------------------------ #
# Validation helpers                                                 #
# ------------------------------------------------------------------ #


def validate_coding_agent_input(input_data: dict) -> CodingAgentInput:
    """
    Validate and construct a CodingAgentInput from a dictionary.

    Raises ValueError if required fields are missing.
    """
    required = {"repo", "branch", "target_branch", "build_command", "test_command"}
    missing = required - set(input_data.keys())
    if missing:
        raise ValueError(f"Missing required fields: {', '.join(sorted(missing))}")

    autonomy_level = input_data.get("autonomy_level")
    if autonomy_level is not None:
        try:
            autonomy_level = AutonomyLevel(autonomy_level)
        except ValueError:
            raise ValueError(
                f"Invalid autonomy_level: {autonomy_level!r}. "
                f"Must be one of: {[a.value for a in AutonomyLevel]}"
            )

    return CodingAgentInput(**input_data)  # type: ignore[arg-type]


def validate_coding_agent_output(output_data: dict) -> CodingAgentOutput:
    """
    Validate and construct a CodingAgentOutput from a dictionary.

    Raises ValueError if required fields are missing.
    """
    required = {"diff", "build_status", "test_results", "lint_results", "recommendation"}
    missing = required - set(output_data.keys())
    if missing:
        raise ValueError(f"Missing required fields: {', '.join(sorted(missing))}")

    recommendation = output_data.get("recommendation")
    if recommendation is not None:
        try:
            recommendation = Recommendation(recommendation)
        except ValueError:
            raise ValueError(
                f"Invalid recommendation: {recommendation!r}. "
                f"Must be one of: {[r.value for r in Recommendation]}"
            )

    return CodingAgentOutput(**output_data)  # type: ignore[arg-type]
